karoo_gp/engine.py

```diff
@@ -37,28 +38,94 @@ def _type_check(self, trees):
         if not all(isinstance(tree, Tree) for tree in trees):
             raise TypeError(f"trees must be a sequence of Trees")
 
+    def parse_expr(self, expr, X_dict, shape):
+        """Parse an expr into a function and insert sample X_dict"""
+        tree = ast.parse(expr, mode='eval').body
+        return self.parse_node(tree, X_dict, shape)
+
+    def parse_node(self, node, X_dict, shape):
+        """Recursively build a function from ast tree and sample X_dict"""
+        if isinstance(node, ast.IfExp):
+            return self.operators['if'](
+                self.parse_node(node.test, X_dict, shape),
+                self.parse_node(node.body, X_dict, shape),
+                self.parse_node(node.orelse, X_dict, shape))
+        if isinstance(node, ast.Name):
+            return X_dict[node.id]
+        elif isinstance(node, ast.Num):
+            return self.num_obj(shape, node.n)
+        elif isinstance(node, ast.UnaryOp):
+            return self.operators[type(node.op)](
+                self.parse_node(node.operand, X_dict, shape))
+        elif isinstance(node, ast.BinOp):
+            return self.operators[type(node.op)](
+                self.parse_node(node.left, X_dict, shape),
+                self.parse_node(node.right, X_dict, shape))
+        elif isinstance(node, ast.BoolOp):
+            return self.operators[type(node.op)](
+                *[self.parse_node(v, X_dict, shape) for v in node.values])
+        elif isinstance(node, ast.Compare):
+            return self.operators[type(node.ops[0])](
+                self.parse_node(node.left, X_dict, shape),
+                self.parse_node(node.comparators[0], X_dict, shape))
+        elif isinstance(node, ast.Call):
+            return self.operators[node.func.id.lower()](
+                self.parse_node(node.args[0], X_dict, shape))
+
+
 #++++++++++++++++++++++++++++
 #   Numpy                   |
 #++++++++++++++++++++++++++++
 
-def inf_to_zero_divide(a, b):
-    # This continued to raise 'RuntimeWarning: divide by zero' errors. This
-    # is a proposed solution. ref: https://stackoverflow.com/a/64747978
-    with np.errstate(divide='ignore'):
-        out = np.where(b==0, 0., a / b)
-    return out
+def safe_divide(a, b):
+    """If dividing by 0, return 0"""
+    return np.where(b==0, 0, a/b)
+
+def safe_sqrt(a):
+    """For sqrt(-a), return -sqrt(abs(a))"""
+    negative = np.less(a, 0)
+    absolute = np.abs(a)
+    square_root = np.sqrt(absolute)
+    square_root[negative] *= -1
```

Comment:
Ah - I think it looks confusing because it's operating on arrays, not values. And with numpy you can index an array using a boolean array. Try the above, but replace `a = -5` with `a = [2, 1, 0, -1, -2]`.

Reply:
With arrays it makes sense.  Maybe it should be clarified in the docstring?

Reply:
Type hints?
```
def safe_divide(a: np.ndarray, b: np.ndarray)
```